src/ride/track_design.c

```diff
@@ -119,187 +115,257 @@ rct_track_td6 *track_design_open(const utf8 *path)
 {
-	rct_track_td6 *td6 = calloc(1, sizeof(rct_track_td6));
-	uint8 *readPtr = src;
-
-	// Read start of track_design
-	copy(td6, &readPtr, 32);
+	rct_track_td4 * td4 = calloc(1, sizeof(rct_track_td4));
+	if (td4 == NULL) {
+		log_error("Unable to allocate memory for TD4 data.");
+		SafeFree(td4);
+		return NULL;
+	}
 
-	uint8 version = td6->version_and_colour_scheme >> 2;
-	if (version > 2) {
+	uint8 version = (src[7] >> 2) & 2;
+	if (version == 0) {
+		memcpy(td4, src, 0x38);
+		td4->elementsSize = srcLength - 0x38;
+		td4->elements = malloc(td4->elementsSize);
+		if (td4->elements == NULL) {
+			log_error("Unable to allocate memory for TD4 element data.");
+			SafeFree(td4);
+			return NULL;
+		}
+		memcpy(td4->elements, src + 0x38, td4->elementsSize);
+	} else if (version == 1) {
+		memcpy(td4, src, 0xC4);
+		td4->elementsSize = srcLength - 0xC4;
+		td4->elements = malloc(td4->elementsSize);
+		if (td4->elements == NULL) {
+			log_error("Unable to allocate memory for TD4 element data.");
+			SafeFree(td4);
+			return NULL;
+		}
+		memcpy(td4->elements, src + 0xC4, td4->elementsSize);
+	} else {
 		log_error("Unsupported track design.");
-		free(td6);
+		SafeFree(td4);
 		return NULL;
 	}
 
-	// In TD6 there are 32 sets of two byte vehicle colour specifiers
-	// In TD4 there are 12 sets so the remaining 20 need to be read
-	if (version == 2) {
-		copy(&td6->vehicle_colours[12], &readPtr, 40);
+	rct_track_td6 * td6 = calloc(1, sizeof(rct_track_td6));
+	if (td6 == NULL) {
+		log_error("Unable to allocate memory for TD6 data.");
+		SafeFree(td4);
+		return NULL;
+	}
+	
+	// Convert RCT1 ride type to RCT2 ride type
+	if (td4->type == RCT1_RIDE_TYPE_WOODEN_ROLLER_COASTER) {
+		td6->type = RIDE_TYPE_WOODEN_ROLLER_COASTER;
+		td6->ride_mode = td4->mode;
+	} else {
+		td6->type = td4->type;
+		td6->ride_mode = td4->mode;
 	}
 
-	copy(&td6->pad_48, &readPtr, 24);
-
-	// In TD4 (version AA/CF) and TD6 both start actual track data at 0xA3
-	if (version > 0) {
-		copy(&td6->track_spine_colour, &readPtr, version == 1 ? 140 : 67);
+	// All TD4s that use powered launch use the type that doesn't pass the station.
+	if (td4->mode == RCT1_RIDE_MODE_POWERED_LAUNCH) {
+		td6->ride_mode = RIDE_MODE_POWERED_LAUNCH;
 	}
 
-	// Read the actual track data to memory directly after the passed in TD6 struct
-	size_t elementDataLength = srcLength - (readPtr - src);
-	uint8 *elementData = malloc(elementDataLength);
-	if (elementData == NULL) {
-		log_error("Unable to allocate memory for TD6 element data.");
-		free(td6);
-		return NULL;
+	// Convert RCT1 vehicle type to RCT2 vehicle type
+	rct_object_entry vehicleObject = { 0x80, { "        " }, 0 };
+	if (td4->type == RIDE_TYPE_MAZE) {
+		const char * name = rct1_get_ride_type_object(td6->type);
+		assert(name != NULL);
+		memcpy(vehicleObject.name, name, min(strlen(name), 8));
+	} else {
+		const char * name = rct1_get_vehicle_object(td6->vehicle_type);
```

Comment:
Should be `td4->vehicle_type`. Changing it solves the issue with the wrong vehicle type.